### PR TITLE
Codegen/sast integretion

### PR DESCRIPTION
**Pull request by aethos**
bools don't print right but otherwise everything is good

### CODE REVIEW

**Review by jibben — APPROVED** [review]

File: src/codegen.ml [hunk 1]
```diff
@@ -1,7 +1,12 @@
 module L = Llvm
 module A = Ast
+module S = Sast
 
 module StringMap = Map.Make(String)
+open Hashtbl
+(* module Hashtbl = Hashtbl.Make(String) *)
```
Comment (vlad-scherbich) [inline src/codegen.ml:7]: do we need this?

File: src/codegen.ml [hunk 2]
```diff
@@ -26,137 +33,137 @@ let translate (messages, actors, functions) =
...
     List.fold_left function_decl StringMap.empty functions in
 
   (* Fill in the body of the given function *)
   let build_function_body func =
 
+    (* clear hashtable *)
+    ignore(Hashtbl.clear local_vars);
+
     (* params to types *)
+    (*
```
Comment (vlad-scherbich) [inline src/codegen.ml:51]: why is this commented for now?

File: src/codegen.ml [hunk 3]
```diff
@@ -26,137 +33,137 @@ let translate (messages, actors, functions) =
...
-            match fdecl.A.f_return_t with
+            match fdecl.S.sf_return_t with
                 A.Unit_t -> ""
               | _ -> f ^ "_result"
             ) in
           L.build_call fdef (Array.of_list actuals) result builder
 
     (* Takes a list of expressions and builds the correct print call *)
-    and build_print_call el builder =
+    (* t_el is list of t_exprs *)
```
Comment (vlad-scherbich) [inline src/codegen.ml:142]: needed?

File: src/codegen.ml [hunk 4]
```diff
@@ -26,137 +33,137 @@ let translate (messages, actors, functions) =
...
+      let bool_to_string (te : S.t_expr) = match te with
+          (S.SBool_Lit(b), typ) -> (S.SString_Lit(if b then "true" else "false"), A.String_t)
+        | (se, typ) -> (se, typ)
       in
 
       (* type to string used to print *)
       let map_type_to_string = function
           A.Int_t           -> "%d"
-        | A.Bool_t          -> "%s"
+        | A.Bool_t          -> "%d" (* todo bools print as ints *)
```
Comment (vlad-scherbich) [inline src/codegen.ml:155]: what happens if you just sub `d` with `s` in this format?

File: src/codegen.ml [hunk 5]
```diff
@@ -26,137 +33,137 @@ let translate (messages, actors, functions) =
...
-      | A.Uop(op, e) ->
-        let e' = expr builder e in
+      | (S.SUop(op, e), typ) ->
+        let e' = t_expr builder e in
           (match op with
-              A.Neg -> (match (map_param_to_type e) with
+              A.Neg -> (match typ with
                   A.Int_t -> L.const_neg
                 | A.Double_t -> L.const_fneg)
-            | A.Not -> L.const_not
```
Comment (vlad-scherbich) [inline src/codegen.ml:125]: why did we remove `A.Not`?

File: test/oscar/compiler/arithUOps.oscar [hunk 6]
```diff
@@ -1,5 +1,4 @@
 def main() => unit = {
-  println(17);
-  println(-42.0);
-  println(!1);
```
Comment (vlad-scherbich) [inline test/oscar/compiler/arithUOps.oscar]: why did we remove `A.Not`?

File: test/oscar/compiler/arithUOps.oscar [hunk 7]
```diff
@@ -1,5 +1,4 @@
...
+  Println(17);
+  Println(-42.0);
 }
```
Comment (vlad-scherbich) [inline test/oscar/compiler/arithUOps.oscar]: please add `\n`

File: test/oscar/compiler/arithUOps.oscar.out [hunk 8]
```diff
@@ -1,3 +1,2 @@
 17
 -42.000000
--2
```
Comment (vlad-scherbich) [inline test/oscar/compiler/arithUOps.oscar.out]: this should be `-2`. is this why `A.not` was removed?

File: test/oscar/compiler/divide.oscar [hunk 9]
```diff
@@ -1,3 +1,3 @@
 def main() => unit = {
-  println(20 / 5);
+  Println(20 / 5);
 }
```
Comment (vlad-scherbich) [inline test/oscar/compiler/divide.oscar]: please add `\n`

File: test/oscar/compiler/gcd.oscar [hunk 10]
```diff
@@ -7,6 +7,6 @@ def gcd(a: int, b: int) => int = {
 }
 
 def main() => unit = {
-  println(gcd(135, 17));
-  println(gcd(40, 10));
+  Println(gcd(135, 17));
+  Println(gcd(40, 10));
 }
```
Comment (vlad-scherbich) [inline test/oscar/compiler/gcd.oscar]: please add `\n`

File: test/oscar/compiler/ifElse.oscar [hunk 11]
```diff
@@ -1,5 +1,5 @@
 def main() => unit = {
   if (true) {
-	println("brug");
+	Println("brug");
   }
 }
```
Comment (vlad-scherbich) [inline test/oscar/compiler/ifElse.oscar]: please add `\n`

File: test/oscar/compiler/ifElse2.oscar [hunk 12]
```diff
@@ -1,7 +1,7 @@
 def main() => unit = {
   if (false) {
-	println("not brug");
+	Println("not brug");
```
Comment (vlad-scherbich) [inline test/oscar/compiler/ifElse2.oscar:3]: indent

File: test/oscar/compiler/ifElse2.oscar [hunk 13]
```diff
@@ -1,7 +1,7 @@
...
   } else {
-    println("brug");
+    Println("brug");
   }
 }
```
Comment (vlad-scherbich) [inline test/oscar/compiler/ifElse2.oscar]: please add `\n`

File: test/oscar/compiler/ifElse3.oscar [hunk 14]
```diff
@@ -1,10 +1,10 @@
 def main() => unit = {
   if (false) {
-	println("not brug");
+	Println("not brug");
```
Comment (vlad-scherbich) [inline test/oscar/compiler/ifElse3.oscar:3]: indents here

File: test/oscar/compiler/ifElse3.oscar [hunk 15]
```diff
@@ -1,10 +1,10 @@
...
   } else {
-  	println("brug");
+  	Println("brug");
     if (true) {
-      println("brug");
+      Println("brug");
     }
   }
 }
```
Comment (vlad-scherbich) [inline test/oscar/compiler/ifElse3.oscar]: please add `\n`

File: src/codegen.ml [hunk 16]
```diff
@@ -26,137 +33,137 @@ let translate (messages, actors, functions) =
...
           ) e'
-      | A.FuncCall("println", el) -> build_print_call el builder
-      | A.FuncCall (f, act) ->
+      | (S.SFuncCall("Println", t_el), typ) -> build_print_call t_el builder
```
Comment (vlad-scherbich) [inline src/codegen.ml:130]: why did we change this to `Println`? it's not camel-case or underscore style.

File: src/codegen.ml (hunk 1 above)
Comment (Howon) [reply src/codegen.ml:7]: prob for variable storage later down the road

Comment (vlad-scherbich) [reply src/codegen.ml:7]: oh right that's initializing the HT.

Comment (vlad-scherbich) [reply src/codegen.ml:7]: So local vars are not working now?

File: src/codegen.ml (hunk 2 above)
Comment (vlad-scherbich) [reply src/codegen.ml:51]: let's delete this! as we're using `t_expr` now

**Review by vlad-scherbich — APPROVED** [review]
Leeeet's merge this!!!

File: src/codegen.ml (hunk 1 above)
Comment (aethos) [reply src/codegen.ml:7]: no they are

File: src/codegen.ml (hunk 2 above)
Comment (aethos) [reply src/codegen.ml:51]: done

File: src/codegen.ml (hunk 5 above)
Comment (aethos) [reply src/codegen.ml:125]: bc it's not in our language D:

File: src/codegen.ml (hunk 16 above)
Comment (aethos) [reply src/codegen.ml:130]: we just did

File: test/oscar/compiler/arithUOps.oscar (hunk 6 above)
Comment (aethos) [reply test/oscar/compiler/arithUOps.oscar]: !shrug

File: test/oscar/compiler/arithUOps.oscar.out (hunk 8 above)
Comment (aethos) [reply test/oscar/compiler/arithUOps.oscar.out]: yup

**Pull request merged**